src/prover.rs

```diff
@@ -425,20 +428,44 @@ where
             )
         });
     let stage_1 = RowMajorMatrix::new(
-        stage_1_on_quotient_domain.vertically_packed_row_pair(i_start, next_step),
+        stage_1_on_quotient_domain.vertically_packed_row_pair::<PackedVal>(i_start, next_step),
         stage_1_width,
     );
+    let extension_d = <PackedExtVal as BasedVectorSpace<PackedVal>>::DIMENSION;
     let stage_2 = RowMajorMatrix::new(
-        stage_2_on_quotient_domain.vertically_packed_row_pair(i_start, next_step),
-        stage_2_width,
+        {
+            let rows = stage_2_on_quotient_domain.wrapping_row_slices(i_start, PackedVal::WIDTH);
+            let next_rows = stage_2_on_quotient_domain
+                .wrapping_row_slices(i_start + next_step, PackedVal::WIDTH);
+
+            (0..stage_2_on_quotient_domain.width())
+                .step_by(extension_d)
+                .map(|c| {
+                    PackedExtVal::from_basis_coefficients_fn(|j| {
+                        PackedVal::from_fn(|i| rows[i][c + j])
+                    })
+                })
+                .chain(
+                    (0..stage_2_on_quotient_domain.width())
+                        .step_by(extension_d)
+                        .map(|c| {
+                            PackedExtVal::from_basis_coefficients_fn(|j| {
+                                PackedVal::from_fn(|i| next_rows[i][c + j])
+                            })
+                        }),
+                )
```

Comment:
Can you wrap these iterators in a generation function that can be called for `rows` and `next_rows`? They seem to have the exact same logic.